File: modules/wem-api/src/main/java/com/enonic/wem/api/content/ContentPermissions.java
package com.enonic.wem.api.content;

import com.google.common.base.Preconditions;

import com.enonic.wem.api.security.acl.AccessControlList;

public final class ContentPermissions
{
    private final AccessControlList contentPermissions;

    private final AccessControlList parentContentPermissions;

    private ContentPermissions( final Builder builder )
    {
        this.contentPermissions = Preconditions.checkNotNull( builder.contentPermissions );
        this.parentContentPermissions = Preconditions.checkNotNull( builder.parentContentPermissions );
    }

    public AccessControlList getPermissions()
    {
        return contentPermissions;
    }

    public AccessControlList getInheritedPermissions()
    {
        return parentContentPermissions;
    }

    public static Builder create()
    {
        return new Builder();
    }

    public static class Builder
    {
        private AccessControlList contentPermissions;

        private AccessControlList parentContentPermissions;

        private Builder()
        {
        }

        public Builder permissions( final AccessControlList value )
        {
            this.contentPermissions = value;
            return this;
        }

        public Builder inheritedPermissions( final AccessControlList value )
        {
            this.parentContentPermissions = value;
            return this;
        }

        public ContentPermissions build()
        {
            return new ContentPermissions( this );
        }
    }

}

File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/ContentPermissions.ts
module api.content {

    export class ContentPermissions {

        private permissions: api.security.acl.AccessControlList;

        private inheritedPermissions: api.security.acl.AccessControlList;

        constructor(builder: ContentPermissionsBuilder) {
            this.permissions = builder.permissions;
            this.inheritedPermissions = builder.inheritedPermissions;
        }


        static create(): ContentPermissionsBuilder {
            return new ContentPermissionsBuilder();
        }

        static fromJson(contentPermissionsJson: api.content.json.ContentPermissionsJson): ContentPermissions {
            return new ContentPermissionsBuilder().fromJson(contentPermissionsJson).build();
        }

    }

    export class ContentPermissionsBuilder {

        permissions: api.security.acl.AccessControlList;

        inheritedPermissions: api.security.acl.AccessControlList;

        constructor() {
            this.permissions = new api.security.acl.AccessControlList();
            this.inheritedPermissions = new api.security.acl.AccessControlList();
        }

        fromJson(contentPermissionsJson: api.content.json.ContentPermissionsJson): ContentPermissionsBuilder {
            this.permissions = api.security.acl.AccessControlList.fromJson(contentPermissionsJson.permissions);
            this.inheritedPermissions = api.security.acl.AccessControlList.fromJson(contentPermissionsJson.inheritedPermissions);
            return this;
        }

        setPermissions(acl: api.security.acl.AccessControlList): ContentPermissionsBuilder {
            this.permissions = acl;
            return this;
        }

        setInheritedPermissions(acl: api.security.acl.AccessControlList): ContentPermissionsBuilder {
            this.inheritedPermissions = acl;
            return this;
        }

        build(): ContentPermissions {
            return new ContentPermissions(this);
        }
    }
}